docs/specification.md

```diff
@@ -0,0 +1,127 @@
+<!-- SPDX-FileCopyrightText: 2020 TBD -->
+
+<!-- SPDX-License-Identifier: LicenseRef-Proprietary -->
+
+# Tezos Globacap
+
+The purpose of this is to create a bridge between Asset Registry
+on Hyperledger to Tezos. This means investors can have immediate liquidity
+of the corresponding investment. This document explains the framework of the
+smart contract.
+
+To restrict and regulate share transfers `Safelist` contract is used.
+
+## Safelist contract
+
+Only admin can access all non-view functions.
+
+Approximate storage structure:
+```haskell
+data Storage = Storage
+  { owner :: Address
+  , admins :: Set Address
+  , whitelist :: Set Address
+  , blacklist :: Set Address
+  }
+```
+
+Required `Safelist` entrypoint:
+* `addAdmin (address)`
+  * Description: adds address to `admins`.
+  * Constraints: `onlyOwner`.
+* `removeAdmin (address)` onlyOwner.
+  * Description: removes address from `admins`.
+  * Constraints: `onlyOwner`.
+* `isAdmin (view address bool)`
+  * Description: returns `True` if the address is admin.
+* `addToWhitelist (address)` onlyAdmin.
+  * Description: adds address to `whitelist`.
+  * Constraints: `onlyAdmin`.
+* `removeFromWhitelist (address)` onlyAdmin.
+  * Description: removes address from `whitelist`.
+  * Constraints: `onlyAdmin`.
+* `isWhitelisted (view address bool)`
+  * Description: returns `True` if the address is whitelisted.
+* `addToBlacklist (address)` onlyAdmin.
+  * Description: adds address to `blacklist`.
+  * Constraints: `onlyAdmin`.
+* `removeFromBlacklist (address)` onlyAdmin.
+  * Description: removes address from `blacklist`.
+  * Constraints: `onlyAdmin`.
+* `isBlacklisted (view address bool)`
+  * Description: returns `True` if the address is blacklisted.
+
+Additionaly, it would be nice to have the following entrypoint:
+* `checkConstraints ([address, constraint])`.
+  * Description: here `constraint` is property that needs to be checked for given
+  address. Can be one of the following:
+  `isWhitelisted`, `isNonBlacklisted`.
```

Comment:
I would specify a specific Michelson type for `constraint` here. It can be `nat` or `int` or `bool` or `or unit unit`.